inmoose/cohort_qc/cohort_metric.py

```diff
@@ -0,0 +1,864 @@
+        """
+        # Perform PCA for data after batch effect correction
+        pca_after = PCA(n_components=n_components)
+        pcs_after = pca_after.fit_transform(self.data_expression_df.T)
+
+        # If data_expression_df_before is not None, compute PCA for it
+        if self.data_expression_df_before is not None:
+            pca_before = PCA(n_components=n_components)
+            pcs_before = pca_before.fit_transform(self.data_expression_df_before.T)
+            association_matrix_before = self._compute_pc_associations(pcs_before)
+        else:
+            pca_before = None
+            pcs_before = None
+            association_matrix_before = None
+
+        # Compute association between PCs and clinical annotations for data after correction
+        association_matrix = self._compute_pc_associations(pcs_after)
+
+        return (
+            association_matrix_before,
+            association_matrix,
+            pca_before,
+            pca_after,
+            pcs_before,
+            pcs_after,
+        )
+
+    def cohort_summary(self) -> dict:
+        """
+        Generate a summary of the cohort, including the number of samples, datasets, genes, and covariate combinations.
+
+        Returns
+        -------
+        dict
+            Summary of the cohort with details on samples, datasets, genes, and covariate combinations.
+        """
+        # Generate a summary of the cohort
+        num_samples = self.clinical_df.shape[0]
+        num_datasets = self.clinical_df[self.batch_column].nunique()
+        num_genes = self.data_expression_df.shape[0]
+        covariates_used = ", ".join(self.covariates)
+
+        # Calculate the number of samples by covariate combinations
+        covariate_combinations = (
+            self.clinical_df.groupby(self.covariates)
+            .size()
+            .reset_index(name="Number of Samples")
+        )
+
+        # Convert the covariate combinations and their counts to a readable format
+        covariate_combination_summary = covariate_combinations.to_dict(orient="records")
+
+        summary = {
+            "Number of Samples": num_samples,
+            "Number of Datasets": num_datasets,
+            "Number of Genes": num_genes,
+            "Covariates Used": covariates_used,
+            "Samples by Covariate Combination": covariate_combination_summary,
+        }
+
+        return summary
+
+    def median_absolute_deviation(self, data: pd.DataFrame) -> np.ndarray:
+        """
+        Compute the median absolute deviation (MAD) for each gene in the data.
+
+        Parameters
+        ----------
+        data : pd.DataFrame
+            Gene expression data frame.
+
+        Returns
+        -------
+        np.ndarray
+            Array of median absolute deviations for each gene.
+        """
+        return np.median(
+            np.abs(data.values - np.median(data.values, axis=1, keepdims=True)), axis=1
+        )
```

Comment:
I think scipy as a function for that, but if results are the same it's ok

Reply:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.median_abs_deviation.html
Indeed, let's not reinvent the wheel 😁

Reply:
Results are the same! i replaced it